src/api/informer.rs

```diff
@@ -126,43 +118,98 @@ impl<K> Informer<K> where
-                *self.version.write().unwrap() = newver;
-                for e in events {
-                    self.events.write().unwrap().push_back(e);
-                }
-            },
+
+        // Create our watch request
+        let req = self.resource.watch(&self.params, &self.version())?;
+
+        // Clone our version so we can move it into the Stream handling
+        // and avoid a 'static lifetime on self
+        let version = self.version.clone();
+
+        // Fetch an initial version as well to use for if we encounter
+        // a desync issue
+        let initial_version = self.get_resource_version().await?;
+
+        // Attempt to fetch our stream
+        let stream = self.client.request_events::<WatchEvent<K>>(req).await;
+
+        match stream {
+            Ok(events) => {
+                // Add a map stage to the stream which will update our version
+                // based on each incoming event
+                Ok(events
+                    .map(move |event| {
+                        // Check if we need to update our version based on the incoming events
+                        let new_version = match &event {
+                            Ok(WatchEvent::Added(o))
+                            | Ok(WatchEvent::Modified(o))
+                            | Ok(WatchEvent::Deleted(o)) => o.meta().resourceVersion.clone(),
+                            Ok(WatchEvent::Error(e)) => {
+                                // Reset our version to the initial version
+                                // and then return a DesyncError.
+                                // This will terminate the stream early in the
+                                // following `take_while`
+                                *version.write().unwrap() = initial_version.clone();
+                                return Err(crate::Error::KubeDesyncError(e.clone()));
+                            }
+                            _ => None,
+                        };
+
+                        // Update our version need be
+                        // Follow docs conventions and store the last resourceVersion
+                        // https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes
+                        if let Some(nv) = new_version {
+                            *version.write().unwrap() = nv;
+                        }
+
+                        event
+                    })
+                    .take_while(|result| {
+                        // If we've run into a KubeDesyncError due to mismatch resource versions
+                        // sleep for 10s before allowing another poll and then terminate the stream early
+                        let err = match result {
+                            Err(crate::Error::KubeDesyncError(e)) => {
+                                warn!("Poll error: {:?}", e);
+                                true
+                            }
+                            _ => false,
```

Comment:
This still causes an error event on the `informer.poll()?` which people usually call in a loop. So it still forces the error handling on users.
I'm tempted to hide this. I.e. expose every WatchEvent::Error, set a flag to resync next poll call, and not expose a DesyncError (maybe log a warning instead - if people want to handle it by scanning, they can catch that Error event in their handler).